lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -0,0 +1,106 @@
+//===- ESIToRTL.cpp - ESI to RTL/SV conversion passes -----------*- C++ -*-===//
+//
+// Lower ESI to RTL and SV.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/ESI/ESIDialect.h"
+#include "circt/Dialect/ESI/ESIOps.h"
+#include "circt/Dialect/RTL/Dialect.h"
+#include "circt/Dialect/SV/Dialect.h"
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/Debug.h"
+
+#include <memory>
+
+namespace circt {
+namespace esi {
+#define GEN_PASS_CLASSES
+#include "circt/Dialect/ESI/ESIPasses.h.inc"
+} // namespace esi
+} // namespace circt
+
+using namespace mlir;
+using namespace circt::esi;
+
+namespace {
+
+/// Lower `ChannelBuffer`s, breaking out the various options. For now, just
+/// replace with the specified number of pipeline stages (since that's the only
+/// option).
+struct ChannelBufferLowering : public ConversionPattern {
```

Comment:
I think this could be `OpConversionPattern<ChannelBuffer>`, and then you wouldn't need to `dyn_cast` the op below.

https://github.com/llvm/llvm-project/blob/c2cb61bed3652126278b4a738e367f524e040ccc/mlir/include/mlir/Transforms/DialectConversion.h#L375